File: engine/engine-model/src/main/scala/net/server/coordinator/partysearch/LeaderSearchMetadata.scala
package net.server.coordinator.partysearch

import client.MapleJob

import scala.util.control._

class LeaderSearchMetadata(private var _minLevel: Int, private var _maxLevel: Int, private var _jobs: Int, private var _jobTable: java.util.Map[Integer, MapleJob]) {
  def minLevel: Int = _minLevel

  def maxLevel: Int = _maxLevel

  def searchedJobs: java.util.List[MapleJob] = _searchedJobs

  private var _reentryCount: Int = 0

  def reentryCount: Int = _reentryCount

  def incrementReentryCount(): Unit = _reentryCount += 1

  private val _searchedJobs: java.util.List[MapleJob] = decodeSearchedJobs(_jobs)

  def decodeSearchedJobs(_jobsSelected: Int): java.util.List[MapleJob] = {
    var jobsSelected = _jobsSelected
    val topByte: Int = ((Math.log(jobsSelected) / Math.log(2)) + 1e-5).asInstanceOf[Int]
    val loop = new Breaks

    loop.breakable {
      for (i <- 0 to topByte) {
        if (jobsSelected % 2 == 1) {
          val job = _jobTable.get(i)
          if (job != null) {
            _searchedJobs.add(job)
          }
          jobsSelected = jobsSelected >> 1
          if (jobsSelected == 0) {
            loop.break()
          }
        }
      }
    }
    _searchedJobs
  }
}

File: engine/engine-model/src/main/java/net/server/coordinator/partysearch/LeaderSearchMetadata.java
package net.server.coordinator.partysearch;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import client.MapleJob;

public class LeaderSearchMetadata {
   private final Integer minLevel;

   private final Integer maxLevel;

   private final Integer jobs;

   private final Map<Integer, MapleJob> jobTable;

   private final List<MapleJob> searchedJobs;

   private Integer reentryCount = 0;

   public LeaderSearchMetadata(Integer minLevel, Integer maxLevel, Integer jobs, Map<Integer, MapleJob> jobTable) {
      this.minLevel = minLevel;
      this.maxLevel = maxLevel;
      this.jobs = jobs;
      this.jobTable = jobTable;
      this.searchedJobs = decodeSearchedJobs(jobs);
   }

   public Integer minLevel() {
      return minLevel;
   }

   public Integer maxLevel() {
      return maxLevel;
   }

   public Integer jobs() {
      return jobs;
   }

   public Map<Integer, MapleJob> jobTable() {
      return jobTable;
   }

   public Integer reentryCount() {
      return reentryCount;
   }

   public void incrementReentryCount() {
      reentryCount += 1;
   }

   public List<MapleJob> searchedJobs() {
      return Collections.unmodifiableList(searchedJobs);
   }

   public List<MapleJob> decodeSearchedJobs(Integer jobsSelected) {
      List<MapleJob> searchedJobs = new LinkedList<>();
      int topByte = (int) ((Math.log(jobsSelected) / Math.log(2)) + 1e-5);
      for (int i = 0; i <= topByte; i++) {
         if (jobsSelected % 2 == 1) {
            MapleJob job = jobTable.get(i);
            if (job != null) {
               searchedJobs.add(job);
            }
         }

         jobsSelected = jobsSelected >> 1;
         if (jobsSelected == 0) {
            break;
         }
      }
      return searchedJobs;
   }
}
